0.11.* coverage plugin causes issues during build deleting sys modules

Hello,

I think the new coverage plugin should be refactored, as it's a potential source of problems.
Looks like it's deleting a bunch of global modules, sparing only the ones that it consider "essential".
We have a custom python installation, the coverage plugin gets a bit confused and considers non essential things like datetime. We end up with errors like:

Traceback (most recent call last):
  File "/tmp/test/env-infra/bin/pyb", line 25, in <module>
    sys.exit(pybuilder.cli.main(*sys.argv[1:]))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 366, in main
    end = datetime.datetime.now()
AttributeError: 'module' object has no attribute 'datetime'


Because of this issue we are unable to update to 0.11.1
I know there should be a way to fix on our side by reverse-engineering your code and put modules in strategic locations, but honestly nothing in pybuilder (nothing in anything IMHO) should play with sys.modules, as it will cause issues that are a nightmare to debug.

Downgrading to 0.10.63 fixes the issue obviously, but the install task would be nice to have for our purposes.

Thanks for the quick reply. Here is the log.
I still think there should be a solution that doesn't require to mess with sys. I did see in the code the part where you walk in the sys lib directory trying to find the "essential" modules, but it still seems like a dangerous approach to me.

PyBuilder version 0.11.1
Build started at 2015-09-23 13:16:04
------------------------------------------------------------
[DEBUG] Loading project module from /tmp/test/build.py
[DEBUG] Loading plugin 'python.core'
[DEBUG] Trying to load builtin plugin 'python.core'
[DEBUG] Loading plugin 'core'
[DEBUG] Trying to load builtin plugin 'core'
[DEBUG] Found builtin plugin 'pybuilder.plugins.core_plugin'
[DEBUG] Found task 'clean' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'clean'
[DEBUG] Found task 'compile_sources' with required dependencies [<function prepare at 0x7f9bb7613c80>] and optional dependencies []
[DEBUG] Registering task 'compile_sources'
[DEBUG] Registering initializer 'init'
[DEBUG] Found task 'install' with required dependencies [<function publish at 0x7f9bb7613f50>] and optional dependencies []
[DEBUG] Registering task 'install'
[DEBUG] Found task 'package' with required dependencies [<function compile_sources at 0x7f9bb7613cf8>] and optional dependencies [<function run_unit_tests at 0x7f9bb7613d70>]
[DEBUG] Registering task 'package'
[DEBUG] Found task 'prepare' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'prepare'
[DEBUG] Found task 'publish' with required dependencies [<function package at 0x7f9bb7613de8>] and optional dependencies [<function verify at 0x7f9bb7613ed8>]
[DEBUG] Registering task 'publish'
[DEBUG] Found task 'run_integration_tests' with required dependencies [<function package at 0x7f9bb7613de8>] and optional dependencies []
[DEBUG] Registering task 'run_integration_tests'
[DEBUG] Found task 'run_unit_tests' with required dependencies [<function compile_sources at 0x7f9bb7613cf8>] and optional dependencies []
[DEBUG] Registering task 'run_unit_tests'
[DEBUG] Found task 'verify' with required dependencies [<function run_integration_tests at 0x7f9bb7613e60>] and optional dependencies []
[DEBUG] Registering task 'verify'
[DEBUG] Found builtin plugin 'pybuilder.plugins.python.core_plugin'
[DEBUG] Registering initializer 'init_python_directories'
[DEBUG] Found task 'package' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'package'
[DEBUG] Loading plugin 'python.unittest'
[DEBUG] Trying to load builtin plugin 'python.unittest'
[DEBUG] Found builtin plugin 'pybuilder.plugins.python.unittest_plugin'
[DEBUG] Registering initializer 'init_test_source_directory'
[DEBUG] Found task 'run_unit_tests' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'run_unit_tests'
[DEBUG] Loading plugin 'python.install_dependencies'
[DEBUG] Trying to load builtin plugin 'python.install_dependencies'
[DEBUG] Found builtin plugin 'pybuilder.plugins.python.install_dependencies_plugin'
[DEBUG] Found action check_pip_available
[DEBUG] Registering action 'check_pip_available'
[DEBUG] Found action create_install_log_directory
[DEBUG] Registering action 'create_install_log_directory'
[DEBUG] Registering initializer 'initialize_install_dependencies_plugin'
[DEBUG] Found task 'install_build_dependencies' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'install_build_dependencies'
[DEBUG] Found task 'install_dependencies' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'install_dependencies'
[DEBUG] Found task 'install_runtime_dependencies' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'install_runtime_dependencies'
[DEBUG] Found task 'list_dependencies' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'list_dependencies'
[DEBUG] Loading plugin 'python.flake8'
[DEBUG] Trying to load builtin plugin 'python.flake8'
[DEBUG] Found builtin plugin 'pybuilder.plugins.python.flake8_plugin'
[DEBUG] Found task 'analyze' with required dependencies ['prepare'] and optional dependencies []
[DEBUG] Registering task 'analyze'
[DEBUG] Found action assert_flake8_is_executable
[DEBUG] Registering action 'assert_flake8_is_executable'
[DEBUG] Registering initializer 'initialize_flake8_plugin'
[DEBUG] Loading plugin 'python.coverage'
[DEBUG] Trying to load builtin plugin 'python.coverage'
[DEBUG] Loading plugin 'analysis'
[DEBUG] Trying to load builtin plugin 'analysis'
[DEBUG] Found builtin plugin 'pybuilder.plugins.analysis_plugin'
[DEBUG] Found task 'analyze' with required dependencies ['run_unit_tests'] and optional dependencies []
[DEBUG] Registering task 'analyze'
[DEBUG] Found builtin plugin 'pybuilder.plugins.python.coverage_plugin'
[DEBUG] Registering initializer 'init_coverage_properties'
[DEBUG] Found action verify_coverage
[DEBUG] Registering action 'verify_coverage'
[DEBUG] Loading plugin 'python.distutils'
[DEBUG] Trying to load builtin plugin 'python.distutils'
[DEBUG] Found builtin plugin 'pybuilder.plugins.python.distutils_plugin'
[DEBUG] Found action build_binary_distribution
[DEBUG] Registering action 'build_binary_distribution'
[DEBUG] Registering initializer 'initialize_distutils_plugin'
[DEBUG] Found task 'install' with required dependencies [] and optional dependencies []
[DEBUG] Registering task 'install'
[DEBUG] Found action write_manifest_file
[DEBUG] Registering action 'write_manifest_file'
[DEBUG] Found action write_setup_script
[DEBUG] Registering action 'write_setup_script'
[DEBUG] Have loaded plugins python.core, core, python.unittest, python.install_dependencies, python.flake8, python.coverage, analysis, python.distutils
[DEBUG] Registering initializer 'set_properties'
[DEBUG] Adding 'prepare' as a required dependency of task 'compile_sources'
[DEBUG] Adding 'publish' as a required dependency of task 'install'
[DEBUG] Adding 'compile_sources' as a required dependency of task 'package'
[DEBUG] Adding 'run_unit_tests' as an optional dependency of task 'package'
[DEBUG] Adding 'package' as a required dependency of task 'publish'
[DEBUG] Adding 'verify' as an optional dependency of task 'publish'
[DEBUG] Adding 'package' as a required dependency of task 'run_integration_tests'
[DEBUG] Adding 'compile_sources' as a required dependency of task 'run_unit_tests'
[DEBUG] Adding 'run_integration_tests' as a required dependency of task 'verify'
[DEBUG] Adding 'prepare' as a required dependency of task 'analyze'
[DEBUG] Adding 'run_unit_tests' as a required dependency of task 'analyze'
[DEBUG] Adding after action 'check_pip_available' for task 'prepare'
[DEBUG] Adding before action 'create_install_log_directory' for task 'install_build_dependencies'
[DEBUG] Adding before action 'create_install_log_directory' for task 'install_runtime_dependencies'
[DEBUG] Adding before action 'create_install_log_directory' for task 'install_dependencies'
[DEBUG] Adding after action 'assert_flake8_is_executable' for task 'prepare'
[DEBUG] Adding after action 'verify_coverage' for task 'analyze'
[DEBUG] Adding after action 'verify_coverage' for task 'verify'
[DEBUG] Adding before action 'build_binary_distribution' for task 'publish'
[DEBUG] Adding after action 'write_manifest_file' for task 'package'
[DEBUG] Adding after action 'write_setup_script' for task 'package'
[DEBUG] Verbose output enabled.

[DEBUG] Executing initializer 'init' from 'pybuilder.plugins.core_plugin'
[DEBUG] Executing initializer 'init_python_directories' from 'pybuilder.plugins.python.core_plugin'
[DEBUG] Executing initializer 'init_test_source_directory' from 'pybuilder.plugins.python.unittest_plugin'
[DEBUG] Executing initializer 'initialize_install_dependencies_plugin' from 'pybuilder.plugins.python.install_dependencies_plugin'
[DEBUG] Executing initializer 'initialize_flake8_plugin' from 'pybuilder.plugins.python.flake8_plugin'
[DEBUG] Executing initializer 'init_coverage_properties' from 'pybuilder.plugins.python.coverage_plugin'
[DEBUG] Executing initializer 'initialize_distutils_plugin' from 'pybuilder.plugins.python.distutils_plugin'
[DEBUG] Executing initializer 'set_properties' from 'build'
[DEBUG] Project properties: 
                                 basedir : /tmp/test
                    coverage_break_build : True
                     coverage_exceptions : []
                           coverage_fork : None
                 coverage_reload_modules : None
                 coverage_threshold_warn : 70
                                dir_dist : $dir_target/dist/test-1.0.dev0
                        dir_dist_scripts : scripts
                        dir_install_logs : $dir_logs/install_dependencies
                                dir_logs : $dir_target/logs
                             dir_reports : $dir_target/reports
                  dir_source_main_python : src/main/python
                 dir_source_main_scripts : src/main/scripts
              dir_source_unittest_python : src/unittest/python
                              dir_target : target
                   distutils_classifiers : ['Development Status :: 3 - Alpha', 'Programming Language :: Python']
                      distutils_commands : ['sdist', 'bdist_dumb']
  distutils_issue8876_workaround_enabled : False
                distutils_use_setuptools : True
                      flake8_break_build : False
                 flake8_exclude_patterns : None
                  flake8_include_scripts : False
             flake8_include_test_sources : False
                  flake8_max_line_length : 120
    install_dependencies_extra_index_url : None
          install_dependencies_index_url : None
      install_dependencies_local_mapping : {}
            install_dependencies_upgrade : False
                    unittest_file_suffix : None
                    unittest_module_glob : *_tests
                         unittest_runner : (<function <lambda> at 0x7f9bb7378050>, u'_make_result')
             unittest_test_method_prefix : None
                                 verbose : True
[DEBUG] Execution plan is prepare, compile_sources, run_unit_tests, package, run_integration_tests, verify, publish
[INFO]  Building test version 1.0.dev0
[INFO]  Executing build in /tmp/test
[INFO]  Going to execute task publish
[DEBUG] Executing task 'prepare'
[DEBUG] Executing subtask from pybuilder.plugins.core_plugin
[DEBUG] Executing action 'check_pip_available' from 'pybuilder.plugins.python.install_dependencies_plugin' before task
[DEBUG] Checking if pip is available
[DEBUG] Executing action 'assert_flake8_is_executable' from 'pybuilder.plugins.python.flake8_plugin' before task
[DEBUG] Checking if flake8 is executable.
[DEBUG] Executing task 'compile_sources'
[DEBUG] Executing subtask from pybuilder.plugins.core_plugin
[DEBUG] Executing task 'run_unit_tests'
[DEBUG] Executing subtask from pybuilder.plugins.core_plugin
[DEBUG] Executing subtask from pybuilder.plugins.python.unittest_plugin
[INFO]  Running unit tests
[DEBUG] Forking process to run unit tests
[INFO]  Executing unit tests from Python modules in /tmp/test/src/unittest/python
[DEBUG] Including files matching '*_tests'
[WARN]  No unit tests executed.
[INFO]  All unit tests passed.
[DEBUG] Executing task 'package'
[DEBUG] Executing subtask from pybuilder.plugins.core_plugin
[DEBUG] Executing subtask from pybuilder.plugins.python.core_plugin
[DEBUG] Removing preexisting distribution /tmp/test/target/dist/test-1.0.dev0
[DEBUG] Creating directory /tmp/test/target/dist/test-1.0.dev0
[INFO]  Building distribution in /tmp/test/target/dist/test-1.0.dev0
[INFO]  Copying scripts to /tmp/test/target/dist/test-1.0.dev0/scripts
[DEBUG] Executing action 'write_manifest_file' from 'pybuilder.plugins.python.distutils_plugin' before task
[DEBUG] No data to write into MANIFEST.in
[DEBUG] Executing action 'write_setup_script' from 'pybuilder.plugins.python.distutils_plugin' before task
[INFO]  Writing setup.py as /tmp/test/target/dist/test-1.0.dev0/setup.py
[DEBUG] Executing task 'run_integration_tests'
[DEBUG] Executing subtask from pybuilder.plugins.core_plugin
[DEBUG] Executing task 'verify'
[DEBUG] Executing subtask from pybuilder.plugins.core_plugin
[DEBUG] Executing action 'verify_coverage' from 'pybuilder.plugins.python.coverage_plugin' before task
[INFO]  Collecting coverage information
[DEBUG] Forking process to do coverage analysis
[DEBUG] Executing task 'prepare'
Traceback (most recent call last):
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 354, in main
    environments=options.environments, tasks=arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/reactor.py", line 163, in build
    reactor=self)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 300, in execute_execution_plan
    summaries.append(self.execute_task(task, **keyword_arguments))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 258, in execute_task
    if self.execute_action(action, keyword_arguments):
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 290, in execute_action
    action.execute(arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 91, in execute
    self.callable(*arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 50, in verify_coverage
    run_coverage(project, logger, reactor, "coverage", "coverage", "run_unit_tests")
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 68, in run_coverage
    target_task, shortest_plan))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 293, in fork_process
    p.start()
  File "/tmp/test/env-infra/lib/python2.7/multiprocessing/process.py", line 130, in start
    self._popen = Popen(self)
  File "/tmp/test/env-infra/lib/python2.7/multiprocessing/forking.py", line 127, in __init__
    sys.stdout.flush()
AttributeError: 'NoneType' object has no attribute 'stdout'
Traceback (most recent call last):
  File "/tmp/test/env-infra/bin/pyb", line 25, in <module>
    sys.exit(pybuilder.cli.main(*sys.argv[1:]))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 366, in main
    end = datetime.datetime.now()
AttributeError: 'module' object has no attribute 'datetime'
Traceback (most recent call last):
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 354, in main
    environments=options.environments, tasks=arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/reactor.py", line 163, in build
    reactor=self)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 300, in execute_execution_plan
    summaries.append(self.execute_task(task, **keyword_arguments))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 258, in execute_task
    if self.execute_action(action, keyword_arguments):
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 290, in execute_action
    action.execute(arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 91, in execute
    self.callable(*arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 50, in verify_coverage
    run_coverage(project, logger, reactor, "coverage", "coverage", "run_unit_tests")
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 68, in run_coverage
    target_task, shortest_plan))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 298, in fork_process
    raise_exception(result[1], result[2])
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 280, in instrumented_target
    send_value = (target(*args, **kwargs), None, None)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 102, in do_coverage
    reactor.execute_task(target_task)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/reactor.py", line 173, in execute_task
    reactor=self)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 300, in execute_execution_plan
    summaries.append(self.execute_task(task, **keyword_arguments))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 228, in execute_task
    timer = Timer.start()
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 204, in start
    return Timer()
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 207, in __init__
    self.start_time = time.time()
AttributeError: 'module' object has no attribute 'time'
------------------------------------------------------------
BUILD FAILED - 'module' object has no attribute 'time'
------------------------------------------------------------
Build finished at 2015-09-23 13:16:05
Build took 0 seconds (811 ms)

I added traceback and apparently sys is also considered not essential. 
I changed the code like this:

    finally:
        import traceback
        traceback.print_exc()
        end = datetime.datetime.now()

and I get
```PyBuilder version 0.11.1
Build started at 2015-09-23 13:28:13
------------------------------------------------------------
[INFO]  Building test version 1.0.dev0
[INFO]  Executing build in /tmp/test
[INFO]  Going to execute task publish
[INFO]  Running unit tests
[INFO]  Executing unit tests from Python modules in /tmp/test/src/unittest/python
[WARN]  No unit tests executed.
[INFO]  All unit tests passed.
[INFO]  Building distribution in /tmp/test/target/dist/test-1.0.dev0
[INFO]  Copying scripts to /tmp/test/target/dist/test-1.0.dev0/scripts
[INFO]  Writing setup.py as /tmp/test/target/dist/test-1.0.dev0/setup.py
[INFO]  Collecting coverage information
Traceback (most recent call last):
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 354, in main
    environments=options.environments, tasks=arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/reactor.py", line 163, in build
    reactor=self)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 300, in execute_execution_plan
    summaries.append(self.execute_task(task, **keyword_arguments))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 258, in execute_task
    if self.execute_action(action, keyword_arguments):
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 290, in execute_action
    action.execute(arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 91, in execute
    self.callable(*arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 50, in verify_coverage
    run_coverage(project, logger, reactor, "coverage", "coverage", "run_unit_tests")
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 68, in run_coverage
    target_task, shortest_plan))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 293, in fork_process
    p.start()
  File "/tmp/test/env-infra/lib/python2.7/multiprocessing/process.py", line 130, in start
    self._popen = Popen(self)
  File "/tmp/test/env-infra/lib/python2.7/multiprocessing/forking.py", line 127, in __init__
    sys.stdout.flush()
AttributeError: 'NoneType' object has no attribute 'stdout'
Traceback (most recent call last):
  File "/tmp/test/env-infra/bin/pyb", line 25, in <module>
    sys.exit(pybuilder.cli.main(*sys.argv[1:]))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 370, in main
    end = datetime.datetime.now()
AttributeError: 'module' object has no attribute 'datetime'
Traceback (most recent call last):
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 354, in main
    environments=options.environments, tasks=arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/reactor.py", line 163, in build
    reactor=self)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 300, in execute_execution_plan
    summaries.append(self.execute_task(task, **keyword_arguments))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 258, in execute_task
    if self.execute_action(action, keyword_arguments):
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 290, in execute_action
    action.execute(arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 91, in execute
    self.callable(*arguments)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 50, in verify_coverage
    run_coverage(project, logger, reactor, "coverage", "coverage", "run_unit_tests")
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 68, in run_coverage
    target_task, shortest_plan))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 298, in fork_process
    raise_exception(result[1], result[2])
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 280, in instrumented_target
    send_value = (target(*args, **kwargs), None, None)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/plugins/python/coverage_plugin.py", line 102, in do_coverage
    reactor.execute_task(target_task)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/reactor.py", line 173, in execute_task
    reactor=self)
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 300, in execute_execution_plan
    summaries.append(self.execute_task(task, **keyword_arguments))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/execution.py", line 228, in execute_task
    timer = Timer.start()
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 204, in start
    return Timer()
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/utils.py", line 207, in __init__
    self.start_time = time.time()
AttributeError: 'module' object has no attribute 'time'
------------------------------------------------------------
BUILD FAILED - 'module' object has no attribute 'time'
------------------------------------------------------------
Build finished at 2015-09-23 13:28:13
Build took 0 seconds (338 ms)
```

Hi, I actually get this:

Collecting https://github.com/pybuilder/pybuilder/archive/issue_211_test.tar.gz
  Downloading https://github.com/pybuilder/pybuilder/archive/issue_211_test.tar.gz (121kB)
    100% |################################| 122kB 460kB/s 
    PyBuilder version ${version}
    Build started at 2015-09-25 10:52:20
    ------------------------------------------------------------
    ------------------------------------------------------------
    BUILD FAILED - Cannot determine git revision: git rev-list HEAD failed:
    fatal: Not a git repository (or any parent up to mount point /tmp)
    Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
    ------------------------------------------------------------
    Build finished at 2015-09-25 10:52:20
    Build took 0 seconds (74 ms)
    Complete output from command python setup.py egg_info:
    PyBuilder version ${version}
    
    Build started at 2015-09-25 10:52:20
    
    ------------------------------------------------------------
    
    ------------------------------------------------------------
    
    BUILD FAILED - Cannot determine git revision: git rev-list HEAD failed:
    
    fatal: Not a git repository (or any parent up to mount point /tmp)
    
    Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
    
    
    
    ------------------------------------------------------------
    
    Build finished at 2015-09-25 10:52:20
    
    Build took 0 seconds (74 ms)
    
    ----------------------------------------
    Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-W2cega-build

Thanks, it installs now! I'm getting this:

```
$ pyb -v
PyBuilder version 0.11.2.dev0
Build started at 2015-09-25 11:09:44
------------------------------------------------------------
[INFO]  Building test version 1.0.dev0
[INFO]  Executing build in /tmp/test
[INFO]  Going to execute task publish
[INFO]  Running unit tests
[INFO]  Executing unit tests from Python modules in /tmp/test/src/unittest/python
[WARN]  No unit tests executed.
[INFO]  All unit tests passed.
[INFO]  Building distribution in /tmp/test/target/dist/test-1.0.dev0
[INFO]  Copying scripts to /tmp/test/target/dist/test-1.0.dev0/scripts
[INFO]  Writing setup.py as /tmp/test/target/dist/test-1.0.dev0/setup.py
[INFO]  Collecting coverage information
[WARN]  coverage_branch_threshold_warn is 0 and branch coverage will not be checked
[WARN]  coverage_branch_partial_threshold_warn is 0 and partial branch coverage will not be checked
Original sys path: ['/tmp/666/env-infra/bin', '/apps/infrafs1/mrossini/.local/lib/python2.7/site-packages/rsa-3.1.2-py2.7.egg', '/apps/infrafs1/mrossini/.local/lib/python2.7/site-packages/pyasn1-0.1.7-py2.7.egg', '/tmp/666/env-infra/lib/python27.zip', '/tmp/666/env-infra/lib/python2.7', '/tmp/666/env-infra/python/2.7.6/lib/python2.7/plat-linux2', '/tmp/666/env-infra/python/2.7.6/lib/python2.7/lib-tk', '/tmp/666/env-infra/lib/python2.7/lib-old', '/tmp/666/env-infra/python/2.7.6/lib/python2.7/lib-dynload', '/apps/infrafs1/mrossini/.local/lib/python2.7/site-packages', '/tmp/666/env-infra/lib/python2.7/site-packages']
Stdlib path: /tmp/666/env-infra/lib/python2.7
Stdlib sys path: ['/tmp/666/env-infra/lib/python2.7', '/tmp/666/env-infra/lib/python2.7/lib-old', '/tmp/666/env-infra/lib/python2.7/site-packages']
Essential (sys module) module 'heapq'
Non-essential module 'distutils'
Essential (sys module) module 'functools'
Essential (sys module) module 'random'
Non-essential module 'datetime'
Essential (sys module) module 'sysconfig'
Essential (built-in) module 'gc'
Non-essential module 'distutils.sysconfig'
Non-essential module 'ctypes._endian'
Essential (sys module) module 'fnmatch'
Essential (sys module) module 'struct'
Essential (sys module) module 'tempfile'
Essential (sys module) module 'pprint'
Non-essential module '_json'
Essential (built-in) module 'imp'
Essential (in sys_packages) module 'pybuilder.plugins.python.setuptools_plugin_helper'
Non-essential module 'multiprocessing.forking'
Essential (in sys_packages) module 'pybuilder.pluginhelper'
Essential (built-in) module 'zipimport'
Essential (sys module) module 'string'
Non-essential module 'multiprocessing.synchronize'
Non-essential module 'encodings.utf_8'
Non-essential module 'unittest.util'
Essential (in sys_packages) module 'pybuilder.plugins.python.coverage_plugin'
Non-essential module 'json.encoder'
Essential (sys module) module 'subprocess'
Essential (built-in) module 'signal'
Essential (in sys_packages) module 'pybuilder'
Essential (sys module) module 'threading'
Essential (sys module) module 'token'
Essential (in sys_packages) module 'pybuilder.ci_server_interaction'
Non-essential module 'multiprocessing'
Essential (sys module) module 'dis'
Non-essential module 'cStringIO'
Essential (sys module) module 'locale'
Essential (in sys_packages) module 'pybuilder.execution'
Non-essential module 'multiprocessing.process'
Essential (sys module) module 'atexit'
Non-essential module 'encodings'
Essential (in sys_packages) module 'pybuilder.plugins.python.flake8_plugin'
Essential (sys module) module 'abc'
Essential (in sys_packages) module 'pybuilder.excp_util_2'
Essential (in sys_packages) module 'pybuilder.errors'
Non-essential module '_multiprocessing'
Essential (sys module) module 're'
Essential (in sys_packages) module 'pybuilder.terminal'
Non-essential module 'unittest.suite'
Non-essential module 'math'
Non-essential module 'fcntl'
Non-essential module 'unittest.case'
Essential (sys module) module 'optparse'
Essential (sys module) module 'UserDict'
Essential (in sys_packages) module 'pybuilder.plugins.core_plugin'
Essential (sys module) module 'inspect'
Essential (in sys_packages) module 'pybuilder.plugins.python.unittest_plugin'
Essential (sys module) module 'Queue'
Non-essential module 'json.scanner'
Non-essential module 'ctypes'
Non-essential module '_ctypes'
Essential (sys module) module 'codecs'
Non-essential module 'unittest.runner'
Non-essential module 'unittest.signals'
Non-essential module 'unittest.loader'
Non-essential module '_functools'
Non-essential module '_locale'
Essential (built-in) module 'thread'
Essential (sys module) module 'StringIO'
Essential (sys module) module 'traceback'
Essential (sys module) module 'weakref'
Non-essential module 'itertools'
Essential (sys module) module 'opcode'
Essential (in sys_packages) module 'pybuilder.core'
Non-essential module 'multiprocessing.util'
Essential (sys module) module 'os'
Essential (built-in) module 'marshal'
Essential (sys module) module '__future__'
Essential (in sys_packages) module 'pybuilder.plugins.python.distutils_plugin'
Non-essential module '_collections'
Essential (built-in) module '_sre'
Non-essential module 'unittest'
Essential (in sys_packages) module 'pybuilder.utils'
Essential (built-in) module '__builtin__'
Essential (in sys_packages) module 'pybuilder.pluginhelper.external_command'
Essential (in sys_packages) module 'pybuilder.plugins.python.core_plugin'
Essential (sys module) module 'platform'
Non-essential module 'operator'
Non-essential module 'select'
Non-essential module '_heapq'
Essential (sys module) module 'posixpath'
Essential (in sys_packages) module 'tblib.cpython'
Essential (built-in) module 'errno'
Non-essential module 'multiprocessing.connection'
Non-essential module '_socket'
Non-essential module 'binascii'
Essential (sys module) module 'sre_constants'
Non-essential module 'json'
Essential (sys module) module 'posixpath'
Essential (sys module) module 'tokenize'
Essential (sys module) module 'collections'
Essential (built-in) module '_warnings'
Non-essential module 'cPickle'
Essential (built-in) module '_codecs'
Essential (in sys_packages) module 'pybuilder.scaffolding'
Essential (in sys_packages) module 'tblib.pickling_support'
Essential (built-in) module 'pwd'
Essential (sys module) module 'copy'
Non-essential module 'unittest.main'
Essential (sys module) module '_sysconfigdata'
Non-essential module '_struct'
Essential (in sys_packages) module 'pybuilder.plugins.python'
Essential (sys module) module 'hashlib'
Essential (in sys_packages) module 'tblib'
Essential (sys module) module 'keyword'
Essential (in sys_packages) module 'pybuilder.cli'
Essential (built-in) module 'posix'
Non-essential module 'encodings.aliases'
Essential (in sys_packages) module 'pybuilder.reactor'
Essential (built-in) module 'exceptions'
Essential (sys module) module 'sre_parse'
Essential (sys module) module 'pickle'
Non-essential module 'json.decoder'
Essential (sys module) module 'copy_reg'
Essential (sys module) module 'sre_compile'
Non-essential module '_hashlib'
Non-essential module '_random'
Essential (sys module) module 'site'
Essential (sys module) module 'io'
Essential (built-in) module '__main__'
Essential (sys module) module 'shutil'
Essential (in sys_packages) module 'pybuilder.plugins.python.install_dependencies_plugin'
Essential (in sys_packages) module 'pybuilder.plugins.analysis_plugin'
Non-essential module 'unittest.result'
Non-essential module 'strop'
Non-essential module 'grp'
Essential (sys module) module 'linecache'
Essential (sys module) module 'socket'
Essential (sys module) module 'gettext'
Non-essential module 'encodings.hex_codec'
Non-essential module 'multiprocessing.queues'
Essential (sys module) module '_abcoll'
Non-essential module 'build'
Essential (sys module) module 'genericpath'
Essential (sys module) module 'stat'
Non-essential module '_ssl'
Essential (sys module) module 'warnings'
Essential (in sys_packages) module 'pybuilder.plugins.python.python_plugin_helper'
Essential (sys module) module 'textwrap'
Essential (built-in) module 'sys'
Essential (in sys_packages) module 'pybuilder.pluginloader'
Essential (sys module) module 'types'
Essential (in sys_packages) module 'pybuilder.graph_utils'
Essential (in sys_packages) module 'pybuilder.plugins'
Essential (built-in) module '_weakref'
Essential (sys module) module 'difflib'
Non-essential module 'distutils.errors'
Non-essential module '_io'
Essential (sys module) module '_weakrefset'
Non-essential module 'time'
Traceback (most recent call last):
  File "/tmp/666/env-infra/bin/pyb", line 25, in <module>
    sys.exit(pybuilder.cli.main(*sys.argv[1:]))
  File "/tmp/666/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 366, in main
    end = datetime.datetime.now()
AttributeError: 'module' object has no attribute 'datetime'
------------------------------------------------------------
BUILD FAILED - 'module' object has no attribute 'time'
------------------------------------------------------------
Build finished at 2015-09-25 11:09:44
Build took 0 seconds (441 ms)
```

Thanks, I can install now! Here is what I get:

```
PyBuilder version 0.11.2.dev0
Build started at 2015-09-25 11:12:08
------------------------------------------------------------
[INFO]  Building test version 1.0.dev0
[INFO]  Executing build in /tmp/test
[INFO]  Going to execute task publish
[INFO]  Running unit tests
[INFO]  Executing unit tests from Python modules in /tmp/test/src/unittest/python
[WARN]  No unit tests executed.
[INFO]  All unit tests passed.
[INFO]  Building distribution in /tmp/test/target/dist/test-1.0.dev0
[INFO]  Copying scripts to /tmp/test/target/dist/test-1.0.dev0/scripts
[INFO]  Writing setup.py as /tmp/test/target/dist/test-1.0.dev0/setup.py
[INFO]  Collecting coverage information
[WARN]  coverage_branch_threshold_warn is 0 and branch coverage will not be checked
[WARN]  coverage_branch_partial_threshold_warn is 0 and partial branch coverage will not be checked
Original sys path: ['/tmp/test/env-infra/bin', '/apps/infrafs1/mrossini/.local/lib/python2.7/site-packages/rsa-3.1.2-py2.7.egg', '/apps/infrafs1/mrossini/.local/lib/python2.7/site-packages/pyasn1-0.1.7-py2.7.egg', '/tmp/test/env-infra/lib/python2.7/site-packages', '/tmp/test/env-infra/lib/python27.zip', '/tmp/test/env-infra/lib/python2.7', '/tmp/test/env-infra/python/2.7.6/lib/python2.7/plat-linux2', '/tmp/test/env-infra/python/2.7.6/lib/python2.7/lib-tk', '/tmp/test/env-infra/lib/python2.7/lib-old', '/tmp/test/env-infra/python/2.7.6/lib/python2.7/lib-dynload', '/apps/infrafs1/mrossini/.local/lib/python2.7/site-packages']
Stdlib path: /tmp/test/env-infra/lib/python2.7
Stdlib sys path: ['/tmp/test/env-infra/lib/python2.7/site-packages', '/tmp/test/env-infra/lib/python2.7', '/tmp/test/env-infra/lib/python2.7/lib-old']
Essential (sys module) module 'heapq'
Non-essential module 'distutils'
Essential (sys module) module 'functools'
Essential (sys module) module 'random'
Non-essential module 'datetime'
Essential (sys module) module 'sysconfig'
Essential (built-in) module 'gc'
Non-essential module 'distutils.sysconfig'
Non-essential module 'ctypes._endian'
Essential (sys module) module 'fnmatch'
Essential (sys module) module 'struct'
Essential (sys module) module 'tempfile'
Essential (sys module) module 'pprint'
Non-essential module '_json'
Essential (built-in) module 'imp'
Essential (in sys_packages) module 'pybuilder.plugins.python.setuptools_plugin_helper'
Non-essential module 'multiprocessing.forking'
Essential (in sys_packages) module 'pybuilder.pluginhelper'
Essential (built-in) module 'zipimport'
Essential (sys module) module 'string'
Non-essential module 'multiprocessing.synchronize'
Non-essential module 'encodings.utf_8'
Non-essential module 'unittest.util'
Essential (in sys_packages) module 'pybuilder.plugins.python.coverage_plugin'
Non-essential module 'json.encoder'
Essential (sys module) module 'subprocess'
Essential (built-in) module 'signal'
Essential (in sys_packages) module 'pybuilder'
Essential (sys module) module 'threading'
Essential (sys module) module 'token'
Essential (in sys_packages) module 'pybuilder.ci_server_interaction'
Non-essential module 'multiprocessing'
Essential (sys module) module 'dis'
Non-essential module 'cStringIO'
Essential (sys module) module 'locale'
Essential (in sys_packages) module 'pybuilder.execution'
Non-essential module 'multiprocessing.process'
Essential (sys module) module 'atexit'
Non-essential module 'encodings'
Essential (in sys_packages) module 'pybuilder.plugins.python.flake8_plugin'
Essential (sys module) module 'abc'
Essential (in sys_packages) module 'pybuilder.excp_util_2'
Essential (in sys_packages) module 'pybuilder.errors'
Non-essential module '_multiprocessing'
Essential (sys module) module 're'
Essential (in sys_packages) module 'pybuilder.terminal'
Non-essential module 'unittest.suite'
Non-essential module 'math'
Non-essential module 'fcntl'
Non-essential module 'unittest.case'
Essential (sys module) module 'optparse'
Essential (sys module) module 'UserDict'
Essential (in sys_packages) module 'pybuilder.plugins.core_plugin'
Essential (sys module) module 'inspect'
Essential (in sys_packages) module 'pybuilder.plugins.python.unittest_plugin'
Essential (sys module) module 'Queue'
Non-essential module 'json.scanner'
Non-essential module 'ctypes'
Non-essential module '_ctypes'
Essential (sys module) module 'codecs'
Non-essential module 'unittest.runner'
Non-essential module 'unittest.signals'
Non-essential module 'unittest.loader'
Non-essential module '_functools'
Non-essential module '_locale'
Essential (built-in) module 'thread'
Essential (sys module) module 'StringIO'
Essential (sys module) module 'traceback'
Essential (sys module) module 'weakref'
Non-essential module 'itertools'
Essential (sys module) module 'opcode'
Essential (in sys_packages) module 'pybuilder.core'
Non-essential module 'multiprocessing.util'
Essential (sys module) module 'os'
Essential (built-in) module 'marshal'
Essential (sys module) module '__future__'
Essential (in sys_packages) module 'pybuilder.plugins.python.distutils_plugin'
Non-essential module '_collections'
Essential (built-in) module '_sre'
Non-essential module 'unittest'
Essential (in sys_packages) module 'pybuilder.utils'
Essential (built-in) module '__builtin__'
Essential (in sys_packages) module 'pybuilder.pluginhelper.external_command'
Essential (in sys_packages) module 'pybuilder.plugins.python.core_plugin'
Essential (sys module) module 'platform'
Non-essential module 'operator'
Non-essential module 'select'
Non-essential module '_heapq'
Essential (sys module) module 'posixpath'
Essential (in sys_packages) module 'tblib.cpython'
Essential (built-in) module 'errno'
Non-essential module 'multiprocessing.connection'
Non-essential module '_socket'
Non-essential module 'binascii'
Essential (sys module) module 'sre_constants'
Non-essential module 'json'
Essential (sys module) module 'posixpath'
Essential (sys module) module 'tokenize'
Essential (sys module) module 'collections'
Essential (built-in) module '_warnings'
Non-essential module 'cPickle'
Essential (built-in) module '_codecs'
Essential (in sys_packages) module 'pybuilder.scaffolding'
Essential (in sys_packages) module 'tblib.pickling_support'
Essential (built-in) module 'pwd'
Essential (sys module) module 'copy'
Non-essential module 'unittest.main'
Essential (sys module) module '_sysconfigdata'
Non-essential module '_struct'
Essential (in sys_packages) module 'pybuilder.plugins.python'
Essential (sys module) module 'hashlib'
Essential (in sys_packages) module 'tblib'
Essential (sys module) module 'keyword'
Essential (in sys_packages) module 'pybuilder.cli'
Essential (built-in) module 'posix'
Non-essential module 'encodings.aliases'
Essential (in sys_packages) module 'pybuilder.reactor'
Essential (built-in) module 'exceptions'
Essential (sys module) module 'sre_parse'
Essential (sys module) module 'pickle'
Non-essential module 'json.decoder'
Essential (sys module) module 'copy_reg'
Essential (sys module) module 'sre_compile'
Non-essential module '_hashlib'
Non-essential module '_random'
Essential (sys module) module 'site'
Essential (sys module) module 'io'
Essential (built-in) module '__main__'
Essential (sys module) module 'shutil'
Essential (in sys_packages) module 'pybuilder.plugins.python.install_dependencies_plugin'
Essential (in sys_packages) module 'pybuilder.plugins.analysis_plugin'
Non-essential module 'unittest.result'
Non-essential module 'strop'
Non-essential module 'grp'
Essential (sys module) module 'linecache'
Essential (sys module) module 'socket'
Essential (sys module) module 'gettext'
Non-essential module 'encodings.hex_codec'
Non-essential module 'multiprocessing.queues'
Essential (sys module) module '_abcoll'
Non-essential module 'build'
Essential (sys module) module 'genericpath'
Essential (sys module) module 'stat'
Non-essential module '_ssl'
Essential (sys module) module 'warnings'
Essential (in sys_packages) module 'pybuilder.plugins.python.python_plugin_helper'
Essential (sys module) module 'textwrap'
Essential (built-in) module 'sys'
Essential (in sys_packages) module 'pybuilder.pluginloader'
Essential (sys module) module 'types'
Essential (in sys_packages) module 'pybuilder.graph_utils'
Essential (in sys_packages) module 'pybuilder.plugins'
Essential (built-in) module '_weakref'
Essential (sys module) module 'difflib'
Non-essential module 'distutils.errors'
Non-essential module '_io'
Essential (sys module) module '_weakrefset'
Non-essential module 'time'
Traceback (most recent call last):
  File "/tmp/test/env-infra/bin/pyb", line 25, in <module>
    sys.exit(pybuilder.cli.main(*sys.argv[1:]))
  File "/tmp/test/env-infra/lib/python2.7/site-packages/pybuilder/cli.py", line 366, in main
    end = datetime.datetime.now()
AttributeError: 'module' object has no attribute 'datetime'
------------------------------------------------------------
BUILD FAILED - 'module' object has no attribute 'time'
------------------------------------------------------------
Build finished at 2015-09-25 11:12:09
Build took 0 seconds (569 ms)
```

--- a/build.py
+++ b/build.py
@@ -96,6 +96,7 @@ def initialize(project):
     project.set_property("verbose", True)
 
     project.set_property("coverage_break_build", False)
+    project.set_property("coverage_reset_modules", True)
     project.get_property("coverage_exceptions").append("pybuilder.cli")
     project.get_property("coverage_exceptions").append("pybuilder.plugins.core_plugin")
 

--- a/src/main/python/pybuilder/plugins/python/coverage_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/coverage_plugin.py
@@ -43,6 +43,7 @@ def init_coverage_properties(project):
     project.set_property_if_unset("coverage_branch_partial_threshold_warn", 0)
     project.set_property_if_unset("coverage_break_build", True)
     project.set_property_if_unset("coverage_reload_modules", None)  # deprecated, unused
+    project.set_property_if_unset("coverage_reset_modules", False)
     project.set_property_if_unset("coverage_exceptions", [])
     project.set_property_if_unset("coverage_fork", None)  # deprecated, unused
 
@@ -88,12 +89,13 @@ def do_coverage(project, logger, reactor, execution_prefix, execution_name, targ
     It's best to simple let this method exit and the fork die rather than to try to recover.
     """
     source_tree_path = project.get_property("dir_source_main_python")
+    reset_modules = project.get_property("%s_reset_modules" % execution_prefix)
     module_names = _discover_modules_to_cover(project)
 
     for module_name in module_names:
         logger.debug("Module '%s' coverage to be verified", module_name)
 
-    if not is_windows():
+    if reset_modules and not is_windows():
         _delete_non_essential_modules()
         __import__("pybuilder.plugins.python")  # Reimport self
 
@@ -330,7 +332,7 @@ def _get_system_assets():
     @return: tuple(packages, modules) to ignore
     """
     canon_sys_path = [os.path.realpath(package_dir) for package_dir in sys.path]
-    std_lib = sysconfig.get_python_lib(standard_lib=True)
+    std_lib = os.path.realpath(sysconfig.get_python_lib(standard_lib=True))
     canon_sys_path = [package_dir for package_dir in canon_sys_path if package_dir.startswith(std_lib)]
 
     packages = []

